apps/ejabberd/src/ejabberd_cowboy.erl

```diff
@@ -50,29 +57,35 @@ socket_type() ->
 
 start_listener({Port, IP, tcp}=Listener, Opts) ->
     IPPort = handler(Listener),
-    ChildSpec = {Listener, {?MODULE, start_link, [IPPort]}, transient,
+    ChildSpec = {Listener, {?MODULE, start_link, [#cowboy_state{ref = cowboy_ref(IPPort), opts = Opts}]}, transient,
                  infinity, worker, [?MODULE]},
     {ok, Pid} = supervisor:start_child(ejabberd_listeners, ChildSpec),
     {ok, _} = start_cowboy(IPPort, [{port, Port}, {ip, IP} | Opts]),
     {ok, Pid}.
 
+reload_dispatch(Ref) ->
+    gen_server:cast(Ref, reload_dispatch).
+
 %% @doc gen_server for handling shutdown when started via ejabberd_listener
 -spec start_link(_) -> 'ignore' | {'error',_} | {'ok',pid()}.
-start_link(Ref) ->
-    gen_server:start_link(?MODULE, [Ref], []).
-init(Ref) ->
+start_link(State) ->
+    gen_server:start_link(?MODULE, State, []).
+init(State) ->
     process_flag(trap_exit, true),
-    {ok, Ref}.
-handle_call(_Request, _From, Ref) ->
-    {noreply, Ref}.
-handle_cast(_Request, Ref) ->
-    {noreply, Ref}.
-handle_info(_Info, Ref) ->
-    {noreply, Ref}.
-code_change(_OldVsn, Ref, _Extra) ->
-    {ok, Ref}.
-terminate(_Reason, Ref) ->
-    stop_cowboy(Ref).
+    {ok, State}.
+handle_call(_Request, _From, State) ->
+    {noreply, State}.
+handle_cast(reload_dispatch, #cowboy_state{ref = Ref, opts = Opts} = State) ->
+    reload_dispatch(Ref, Opts),
+    {noreply, State};
+handle_cast(_Request, State) ->
+    {noreply, State}.
+handle_info(_Info, State) ->
+    {noreply, State}.
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+terminate(_Reason, State) ->
+    stop_cowboy(State).
```

Comment:
This ain't gonna work :) `stop_cowboy` takes the cowboy's reference and not process state... We should start using dialyzer better...